File: papyra.api/tests/edge/json_get.js
// Node twin of json_get.py, used when no Python is on PATH.
// Same contract: a dotted path on argv, JSON on stdin, one value on stdout,
// an empty line for anything that does not resolve.

const path = process.argv[2] || '';
let raw = '';

process.stdin.setEncoding('utf8');
process.stdin.on('data', (chunk) => { raw += chunk; });
process.stdin.on('end', () => {
  let doc;
  try {
    doc = JSON.parse(raw);
  } catch {
    process.stdout.write('\n');
    return;
  }

  for (const part of path.split('.')) {
    if (part === '') continue;
    if (doc === null || doc === undefined) break;
    if (/^\d+$/.test(part) && Array.isArray(doc)) {
      doc = doc[Number(part)];
    } else if (typeof doc === 'object') {
      if (part in doc) {
        doc = doc[part];
      } else {
        // Match the Python twin: fall back to a case-insensitive lookup so a
        // check does not have to know which casing an endpoint used.
        const key = Object.keys(doc).find((k) => k.toLowerCase() === part.toLowerCase());
        doc = key === undefined ? undefined : doc[key];
      }
    } else {
      doc = undefined;
    }
  }

  if (doc === null || doc === undefined) process.stdout.write('\n');
  else if (typeof doc === 'object') process.stdout.write(JSON.stringify(doc) + '\n');
  else process.stdout.write(String(doc) + '\n');
});

File: papyra.api/tests/edge/json_get.py
"""Read one value out of a JSON document on stdin.

Usage:  json_get.py <dotted-path>

The path is dotted, with numeric segments meaning "index into a list":

    id            -> doc["id"]
    0.id          -> doc[0]["id"]
    user.name     -> doc["user"]["name"]
    ""            -> the whole document

Anything that cannot be resolved prints an empty line and exits 0, so a
harness check reads a missing field as "" rather than as a crash. Booleans
print as `true`/`false` (not Python's `True`/`False`) so a shell comparison
against the JSON the API actually sent works.
"""

import json
import sys


def resolve(doc, path):
    for part in path.split("."):
        if part == "":
            continue
        if part.isdigit() and isinstance(doc, list):
            index = int(part)
            if index >= len(doc):
                return None
            doc = doc[index]
        elif isinstance(doc, dict):
            # The API serialises camelCase, but a caller writing a check should
            # not have to remember which casing a given endpoint used.
            if part in doc:
                doc = doc[part]
            else:
                lowered = {k.lower(): v for k, v in doc.items()}
                doc = lowered.get(part.lower())
        else:
            return None
        if doc is None:
            return None
    return doc


def main():
    path = sys.argv[1] if len(sys.argv) > 1 else ""
    try:
        doc = json.load(sys.stdin)
    except (ValueError, UnicodeDecodeError):
        print("")
        return
    value = resolve(doc, path)
    if value is None:
        print("")
    elif isinstance(value, bool):
        print("true" if value else "false")
    elif isinstance(value, (dict, list)):
        print(json.dumps(value))
    else:
        print(value)


if __name__ == "__main__":
    main()
